docs(docs/design): `Network` and `Syncing` Implementation Design (#4775)
ChainSafe/gossamer: docs/docs/design/network-syncing.md

---
layout: default
title: Network and Syncing Interface Implementation
permalink: /design/network-syncing-interfaces/
---
# `Network` and `Syncing` Implementation Design

The [`Network`] and [`Syncing`] interfaces were introduced in [PR #4572](https://github.com/ChainSafe/gossamer/pull/4572) as part of the `gossip` package and represent the network and syncing interfaces that are dependencies for the `GossipEngine` type.  What follows is an assessment on each method found in both `Network` and `Syncing` interfaces and how we propose to implement them using components already found in the Gossamer codebase and/or introduce new components.

## [`Network`]

The `Network` interface is as follows:
```go
// Network is the abstraction over a network.
type Network interface {
	service.NetworkPeers
	service.NetworkEventStream

> 	AddSetReserved(who peerid.PeerID, protocol network.ProtocolName)

Reviewer: Why keep them separate when `service.NetworkPeers` contains all other methods for manipulating reserved peer sets?

Author: `AddSetReserved` is a helper function and is implemented on the trait itself.  I think it makes sense to keep it this way for now, and we can refactor the interfaces afterwards in more of a cleanup effort after we validate things are working.

	// have any substream open with them yet.
	//
	// Note however, if a call to this function results in less peers on the reserved set, they will not necessarily
	// get disconnected (depending on available free slots in the peer set). If you want to also disconnect those
	// removed peers, you will have to call RemoveFromPeersSet on those in addition to updating the reserved set. You
	// can omit this step if the peer set is in reserved only mode.
	//
	// Returns an error if one of the given addresses is invalid or contains an invalid peer ID (which includes the
	// local peer id).
	SetReservedPeers(protocol network.ProtocolName, peers map[multiaddr.Multiaddr]struct{}) error
	// Add peers to a peer set.
	//
	// Each Multiaddr must end with a "/p2p/" component containing the peer id. It can also consist of only
	// "/p2p/<peerid>".
	//
	// Returns an error if one of the given addresses is invalid or contains an invalid peer id (which includes the
	// local peer id).
	AddPeersToReservedSet(protocol network.ProtocolName, peers map[multiaddr.Multiaddr]struct{}) error
	// Remove peers from a peer set.
	RemovePeersFromReservedSet(protocol network.ProtocolName, peers []peerid.PeerID)

> 	// Add a peer to a set of peers.

Reviewer: This is for multiple peers, right?
```suggestion
	// Add peers to a set of peers.
```
and change the doc comment on `AddPeersToReservedSet()` to
```
	// Add peers to a reserved peers set.
```
I assume that is the only difference between the two methods.

Author: This was removed in substrate, and I removed it in af8d4724604312d98eddcbeb73b9778078700b7e, which is part of this PR now.

	//
	// If the set has slots available, it will try to open a substream with this peer.
	//
	// Each Multiaddr must end with a "/p2p/" component containing the peer id. It can also consist of only
	// "/p2p/<peerid>".
	//
	// Returns an error if one of the given addresses is invalid or contains an invalid peer id (which includes the
	// local peer id).
	AddToPeersSet(protocol network.ProtocolName, peers map[multiaddr.Multiaddr]struct{}) error
	// Remove peers from a peer set.
	//

> 	// If we currently have an open substream with this peer, it will soon be closed.

Reviewer: ```suggestion
	// If we currently have open substreams with any of these peers, they will soon be closed.
```

	RemoveFromPeersSet(protocol network.ProtocolName, peers []peerid.PeerID)
	// Returns the number of peers in the sync peer set we're connected to.

> 	SyncNumConnected() uint

Reviewer 2: if it is not used anywhere maybe we should remove it from inteface?

Author: I'd like to keep it in for future work, but the intention is to keep it unimplemented in the translation shim.

	// Attempt to get peer role.
	//
	// Right now the peer role is decoded from the received handshake for all protocols ("/block-announces/1" has other

> 	// information as well). If the handshake cannot be decoded into a role, the role queried from peer storeand if the

Reviewer: ```suggestion
	// information as well). If the handshake cannot be decoded into a role, the role is queried from peer store and if the
```

### [`NetworkBlock`]

[`NetworkBlock`] contains two methods, `AnnounceBlock` and `NewBestBlockImported`. 

```go
// NetworkBlock provides ability to announce blocks to the network.
type NetworkBlock[BlockHash any, BlockNumber any] interface {
	// Make sure an important block is propagated to peers.
	//
	// In chain based consensus, we often need to make sure non-best forks are at least temporarily synced. This
	// function forces such an announcement.
	AnnounceBlock(hash BlockHash, data []byte)

	// Inform the network service about new best imported block.
	NewBestBlockImported(hash BlockHash, number BlockNumber)
}

```

`AnnounceBlock` method is used by `GossipEngine` to announce blocks before GRANDPA specific messages are sent.  We already have functionality to create instances `network.BlockAnnounceMessage` using [`core.createBlockAnnounce`](https://github.com/ChainSafe/gossamer/blob/205b72beca190262c12bd13ccd38d12bd026d06f/dot/core/service.go#L230) that can be sent over the `Network` via `Network.GossipMessage`.  We will need to replicate/make public the `BlockAnnounceMessage` constructor that also has the `Network` to send block announces.

> `NewBestBlockImported` is currently not used in the Gossamer codebase and can remain unimplemented.

Reviewer: I wonder whether not sending these negatively affects our nodes reputation.

Author: I don't think so, since this is more of a repeated block announce right before we send the signed GRANDPA protocol messages.  I didn't see anything in the GRANDPA specific reputation changes either ([link](https://github.com/paritytech/polkadot-sdk/blob/d3128dc679a089750a8569acc6e0834787a7b87c/substrate/client/consensus/grandpa/src/communication/mod.rs#L100)).
